src/Compiler/Utilities/Cancellable.fs

```diff
@@ -51,176 +52,248 @@ type Cancellable =
-            let (Cancellable g) = f ()
-            g ct)
+    member inline _.Zero() : CancellableCode<'Data, unit> = ResumableCode.Zero()
 
-    member inline _.Bind(comp, [<InlineIfLambda>] k) =
-        Cancellable(fun ct ->
+    member inline _.For(sequence, body) : CancellableCode<'Data, unit> = ResumableCode.For(sequence, body)
 
-            __debugPoint ""
+    member inline _.While(condition, body) : CancellableCode<'Data, unit> =
+        ResumableCode.While(condition, CancellableCode.WithCancelCheck body)
 
-            match Cancellable.run ct comp with
-            | ValueOrCancelled.Value v1 -> Cancellable.run ct (k v1)
-            | ValueOrCancelled.Cancelled err1 -> ValueOrCancelled.Cancelled err1)
+    member inline _.Delay(generator: unit -> CancellableCode<'Data, 'T>) =
+        CancellableCode<'Data, 'T>(fun sm -> (generator ()).Invoke(&sm))
 
-    member inline _.BindReturn(comp, [<InlineIfLambda>] k) =
-        Cancellable(fun ct ->
+    member inline _.Combine(code1, code2) =
+        ResumableCode.Combine(CancellableCode.WithCancelCheck code1, CancellableCode.WithCancelCheck code2)
 
-            __debugPoint ""
+    member inline _.Using(resource, body) : CancellableCode<'Data, 'T> = ResumableCode.Using(resource, body)
 
-            match Cancellable.run ct comp with
-            | ValueOrCancelled.Value v1 -> ValueOrCancelled.Value(k v1)
-            | ValueOrCancelled.Cancelled err1 -> ValueOrCancelled.Cancelled err1)
+    member inline _.TryWith(body, catch) : CancellableCode<'Data, 'T> =
+        ResumableCode.TryWith(CancellableCode.WithCancelCheck body, (CancellableCode.FilterOce catch))
 
-    member inline _.Combine(comp1, comp2) =
-        Cancellable(fun ct ->
+    member inline _.TryFinally(body: CancellableCode<'Data, 'T>, compensation) : CancellableCode<'Data, 'T> =
+        ResumableCode.TryFinally(
+            CancellableCode.WithCancelCheck body,
+            ResumableCode(fun _sm ->
+                compensation ()
+                true)
+        )
 
-            __debugPoint ""
+    member inline _.Return(value: 'T) : CancellableCode<'T, 'T> =
+        CancellableCode(fun sm ->
+            sm.Data.Result <- Ok value
+            true)
 
-            match Cancellable.run ct comp1 with
-            | ValueOrCancelled.Value() -> Cancellable.run ct comp2
-            | ValueOrCancelled.Cancelled err1 -> ValueOrCancelled.Cancelled err1)
+    member inline this.Yield(value) = this.Return(value)
 
-    member inline _.TryWith(comp, [<InlineIfLambda>] handler) =
-        Cancellable(fun ct ->
+    member inline _.Bind
+        (code: Cancellable<'U>, [<InlineIfLambda>] continuation: 'U -> CancellableCode<'Data, 'T>)
+        : CancellableCode<'Data, 'T> =
+        CancellableCode(fun sm ->
+            if __useResumableCode then
+                let mutable invocation = code.GetInvocation()
 
-            __debugPoint ""
+                let __stack_yield_complete = ResumableCode.Yield().Invoke(&sm)
```

Comment:
Yielding here makes the bound code trampoline. This should not yield all the time but conditionally on bind depth (not yet implemented).